kotlinx-coroutines-test/common/src/TestBuilders.kt

```diff
@@ -269,46 +278,126 @@ public fun runTest(
  *
  * @throws IllegalArgumentException if the [context] is invalid. See the [TestScope] constructor docs for details.
  */
-@ExperimentalCoroutinesApi
+@Deprecated(
+    "Define a total timeout for the whole test instead of using dispatchTimeoutMs. " +
+        "Warning: the proposed replacement is not identical as it uses 'dispatchTimeoutMs' as the timeout for the whole test!",
+    ReplaceWith("runTest(context, timeout = dispatchTimeoutMs.milliseconds, testBody)",
+        "kotlin.time.Duration.Companion.milliseconds"),
+    DeprecationLevel.WARNING
+)
 public fun runTest(
     context: CoroutineContext = EmptyCoroutineContext,
     dispatchTimeoutMs: Long,
     testBody: suspend TestScope.() -> Unit
-): TestResult = runTest(
-    context = context,
-    dispatchTimeout = dispatchTimeoutMs.milliseconds,
-    testBody = testBody
-)
+): TestResult {
+    if (context[RunningInRunTest] != null)
+        throw IllegalStateException("Calls to `runTest` can't be nested. Please read the docs on `TestResult` for details.")
+    @Suppress("DEPRECATION")
+    return TestScope(context + RunningInRunTest).runTest(dispatchTimeoutMs = dispatchTimeoutMs, testBody)
+}
 
 /**
- * Performs [runTest] on an existing [TestScope].
+ * Performs [runTest] on an existing [TestScope]. See the documentation for [runTest] for details.
  */
 @ExperimentalCoroutinesApi
 public fun TestScope.runTest(
-    dispatchTimeout: Duration,
+    timeout: Duration = DEFAULT_TIMEOUT,
     testBody: suspend TestScope.() -> Unit
 ): TestResult = asSpecificImplementation().let {
     it.enter()
     createTestResult {
-        runTestCoroutine(it, dispatchTimeout, TestScopeImpl::tryGetCompletionCause, testBody) {
+        /** TODO: moving this [AbstractCoroutine.start] call outside [createTestResult] fails on JS. */
+        it.start(CoroutineStart.UNDISPATCHED, it) {
+            testBody()
+        }
+        /**
+         * We run the tasks in the test coroutine using [Dispatchers.Default]. On JS, this does nothing particularly,
+         * but on the JVM and Native, this means that the timeout can be processed even while the test runner is busy
+         * doing some synchronous work.
+         */
+        val workRunner = launch(Dispatchers.Default + CoroutineName("kotlinx.coroutines.test runner")) {
```

Comment:
Could you please elaborate a bit on the reasoning behind this change?

From now on, `runTest` behaves in a bit surprising manner -- it changes the execution thread from "main" (where all tests are launched, a single-threaded default) to "worker" and it may be an observable behaviour:
```
    val tl = ThreadLocal.withInitial { 42 }

    @Test
    fun foo() = runTest {
        println(Thread.currentThread())
        tl.set(41)
        println(tl.get())
        yield()
        println(Thread.currentThread())
        println(tl.get())
    }
```

it also may pollute thread-locals arbitrarily (I believe this is the reason `TimeoutRule` creates standalone thread each time).

It may potentially confuse tooling around tests and interferes with production code (less threads in `Dispatchers.Default` that can actually matter on low-number-of-cores-machines, and interferes with thread-locals when used carelessly) for the sake of a more granular timeout.

Maybe it's worth to run this loop in-place (instead of standalone coroutine) and check for timeout cancellation at the end. We are going to wait for syncrhnous work anyway in order to terminate the test, but the interference issues will be addressed

Reply:
The reason is simple: I wanted to get a coroutine dump at the moment when everything's hanged. I always thought `TimeoutRule`, too, spawned a separate thread for this reason. You do raise a good point though that due to `CoroutineStart.UNDISPATCHED`, the threading behavior is surprising.

I see these ways to mitigate the problem:
* Start a separate thread for each test and ensure that the test immediately executes on that thread.
* Do run the loop on the main thread, but replace `withTimeout` with custom code that cancels the test and dumps the coroutines.

Reply:
Thanks!

>Start a separate thread for each test and ensure that the test immediately executes on that thread.

Let's consider this one as the last resort if the second option fails

>but replace withTimeout

Wouldn't it be enough to keep `withTimeout` with `invokeOnCancellation(onCancelling = true) { if (isIsATimeoutWeExpecting) dumpCoroutines() } `?

Reply:
`invokeOnCompletion`? This is not quite enough, because to support cancellation when the test body doesn't suspend and doesn't release the thread, we need to cancel the test coroutine asynchronously. I expanded on the solution you proposed and added a test that shows this problem and tests the thread dumps.